File: AspNetMvcUnleashed/0672329980_source/Chapter06/VB/MvcApplication1/MvcApplication1.Tests/Helpers/DataGridHelperTests.vb
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports MvcFakes
Imports System.Text.RegularExpressions

<TestClass()> _
Public Class DataGridHelperTests

    Public Function CreateItems(ByVal count As Integer) As List(Of Product)
        Dim items = New List(Of Product)()
        For i = 0 To count - 1
            Dim newProduct = New Product()
            newProduct.Id = i
            newProduct.Name = String.Format("Product {0}", i)
            newProduct.Price = count - i
            items.Add(newProduct)
        Next i
        Return items
    End Function


    <TestMethod()> _
    Public Sub SecondPageNumberSelected()
        ' Arrange
        Dim items = CreateItems(5)
        Dim data = items.AsQueryable().ToPagedList(1, 2)

        ' Act
        Dim fakeHtmlHelper = New FakeHtmlHelper()
        Dim results = DataGridHelper.DataGrid(Of Product)(fakeHtmlHelper, data)

        ' Assert
        StringAssert.Contains(results, "<strong>2</strong>")

    End Sub


    <TestMethod()> _
    Public Sub CorrectNumberOfRows()
        ' Arrange
        Dim items = CreateItems(5)
        Dim data = items.AsQueryable().ToPagedList(1, 2)

        ' Act
        Dim fakeHtmlHelper = New FakeHtmlHelper()
        Dim results = DataGridHelper.DataGrid(Of Product)(fakeHtmlHelper, data)

        ' Assert (1 header row + 2 data rows + 1 pager row)
        Assert.AreEqual(4, Regex.Matches(results, "<tr>").Count)
    End Sub
End Class

Public Class Product
    Private _id As Integer
    Private _name As String
    Private _price As Decimal

    Public Property Id() As Integer
        Get
            Return _id
        End Get
        Set(ByVal value As Integer)
            _id = value
        End Set
    End Property

    Public Property Name() As String
        Get
            Return _name
        End Get
        Set(ByVal value As String)
            _name = value
        End Set
    End Property

    Public Property Price() As Decimal
        Get
            Return _price
        End Get
        Set(ByVal value As Decimal)
            _price = value
        End Set
    End Property
End Class

File: AspNetMvcUnleashed/0672329980_source/Chapter06/CS/MvcApplication1/MvcApplication1.Tests/Helpers/DataGridHelperTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using Helpers;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using MvcFakes;
using Paging;

namespace MvcApplication1.Tests.Helpers
{
    [TestClass]
    public class DataGridHelperTests
    {

        public List<Product> CreateItems(int count)
        {
            var items = new List<Product>();
            for (var i=0;i < count;i++)
            {
                var newProduct = new Product();
                newProduct.Id = i;
                newProduct.Name = String.Format("Product {0}", i);
                newProduct.Price = count - i;
                items.Add(newProduct);
            }
            return items;
        }


        [TestMethod]
        public void SecondPageNumberSelected()
        {
            // Arrange
            var items = CreateItems(5);
            var data = items.AsQueryable().ToPagedList(1, 2); 

            // Act
            var fakeHtmlHelper = new FakeHtmlHelper();
            var results = DataGridHelper.DataGrid<Product>(fakeHtmlHelper, data);

            // Assert
            StringAssert.Contains(results, "<strong>2</strong>");
        
        }


        [TestMethod]
        public void CorrectNumberOfRows()
        {
            // Arrange
            var items = CreateItems(5);
            var data = items.AsQueryable().ToPagedList(1, 2);

            // Act
            var fakeHtmlHelper = new FakeHtmlHelper();
            var results = DataGridHelper.DataGrid<Product>(fakeHtmlHelper, data);

            // Assert (1 header row + 2 data rows + 1 pager row)
            Assert.AreEqual(4, Regex.Matches(results, "<tr>").Count);
        }
    }

    public class Product
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public decimal Price { get; set; }
    }
}
